File: app/src/main/java/com/siziksu/services/service/MessengerService.java
package com.siziksu.services.service;

import android.app.Service;
import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.IBinder;
import android.os.Message;
import android.os.Messenger;

import com.siziksu.services.common.Commons;
import com.siziksu.services.common.Constants;
import com.siziksu.services.common.Mock;

public class MessengerService extends Service {

  public static final int MESSENGER_SERVICE_TEST = 0;
  private static final int DELAY_TIME_TO_PUBLISH_REQUEST_RECEIVED = 2000;

  private Messenger messenger = new Messenger(new RequestHandler());

  @Override
  public IBinder onBind(Intent intent) {
    Commons.log(Constants.TAG_MESSENGER_SERVICE, Constants.SERVICE_BOUND);
    return messenger.getBinder();
  }

  @Override
  public boolean onUnbind(Intent intent) {
    Commons.log(Constants.TAG_MESSENGER_SERVICE, Constants.SERVICE_UNBOUND);
    return super.onUnbind(intent);
  }

  @Override
  public void onCreate() {
    super.onCreate();
    Commons.log(Constants.TAG_MESSENGER_SERVICE, Constants.SERVICE_CREATED);
  }

  @Override
  public int onStartCommand(Intent intent, int flags, int startId) {
    Commons.log(Constants.TAG_MESSENGER_SERVICE, Constants.SERVICE_STARTED);
    return START_STICKY;
  }

  @Override
  public boolean stopService(Intent name) {
    Commons.log(Constants.TAG_MESSENGER_SERVICE, Constants.SERVICE_STOPPED);
    return super.stopService(name);
  }

  @Override
  public void onDestroy() {
    super.onDestroy();
    Commons.log(Constants.TAG_MESSENGER_SERVICE, Constants.SERVICE_DESTROYED);
  }

  static class RequestHandler extends Handler {

    public RequestHandler() {}

    @Override
    public void handleMessage(Message msg) {
      switch (msg.what) {
        case MESSENGER_SERVICE_TEST:
          if (msg.getData().containsKey(Constants.EXTRAS_JSON)) {
            String json = msg.getData().getString(Constants.EXTRAS_JSON);
            Commons.log(Constants.TAG_MESSENGER_SERVICE, json);
            Mock.getInstance().pause(DELAY_TIME_TO_PUBLISH_REQUEST_RECEIVED);
            Message response = Message.obtain(null, MESSENGER_SERVICE_TEST);
            Bundle bundle = new Bundle();
            bundle.putString(Constants.EXTRAS_JSON, Mock.getInstance().fakeResponse());
            response.setData(bundle);
            try {
              msg.replyTo.send(response);
            } catch (Exception e) {
              e.printStackTrace();
            }
          }
          break;

        default:
          super.handleMessage(msg);
      }
    }
  }

  public interface Listener {

    void onMessage(String response);
  }
}

File: app/src/main/java/com/siziksu/services/data/service/MessengerService.kt
package com.siziksu.services.data.service

import android.app.Service
import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.os.IBinder
import android.os.Message
import android.os.Messenger
import com.siziksu.services.app.Constants
import com.siziksu.services.commons.Commons
import com.siziksu.services.commons.mock.Mock

class MessengerService : Service() {

    private val messenger = Messenger(RequestHandler())

    override fun onBind(intent: Intent): IBinder? {
        Commons.log(Constants.TAG_MESSENGER_SERVICE, Constants.SERVICE_BOUND)
        return messenger.binder
    }

    override fun onUnbind(intent: Intent): Boolean {
        Commons.log(Constants.TAG_MESSENGER_SERVICE, Constants.SERVICE_UNBOUND)
        return super.onUnbind(intent)
    }

    override fun onCreate() {
        super.onCreate()
        Commons.log(Constants.TAG_MESSENGER_SERVICE, Constants.SERVICE_CREATED)
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        Commons.log(Constants.TAG_MESSENGER_SERVICE, Constants.SERVICE_STARTED)
        return START_STICKY
    }

    override fun stopService(name: Intent): Boolean {
        Commons.log(Constants.TAG_MESSENGER_SERVICE, Constants.SERVICE_STOPPED)
        return super.stopService(name)
    }

    override fun onDestroy() {
        super.onDestroy()
        Commons.log(Constants.TAG_MESSENGER_SERVICE, Constants.SERVICE_DESTROYED)
    }

    internal class RequestHandler : Handler() {

        override fun handleMessage(msg: Message) {
            when (msg.what) {
                MESSENGER_SERVICE_WHAT -> // Message received in this "what"
                    if (msg.data.containsKey(Constants.EXTRAS_JSON)) {
                        val json = msg.data.getString(Constants.EXTRAS_JSON)
                        Commons.log(Constants.TAG_MESSENGER_SERVICE, json ?: "") // Prints the request data

                        val response = Message.obtain(null, MESSENGER_SERVICE_WHAT) // Prepares the message to be sent

                        val bundle = Bundle()
                        bundle.putString(Constants.EXTRAS_JSON, Mock.fakeResponse())
                        response.data = bundle // Sets the data to be sent

                        try {
                            msg.replyTo.send(response) // Sends the data
                        } catch (e: Exception) {
                            Commons.error(e)
                        }
                    }
                else -> super.handleMessage(msg)
            }
        }
    }

    companion object {

        const val MESSENGER_SERVICE_WHAT = 0
    }

    interface Listener {

        fun onMessage(response: String)
    }
}
